src/classes/RD2_ScheduleService.cls

```diff
@@ -319,4 +319,120 @@ public without sharing class RD2_ScheduleService {
+            }
+
+            for (SObject sobj : revisedSchedules.objectsToInsert) {
+                addUpdatedSchedules((RecurringDonationSchedule__c)sobj, newSchedulesByRdId);
+            }
+        }
+
+        // Merge the schedules from RDs NOT needing schedule updates with the updated schedules
+        for (Id rdId : newSchedulesByRdId.keySet()) {
+            schedulesByRdId.put(rdId, newSchedulesByRdId.get(rdId));
+        }
+
+        for (npe03__Recurring_Donation__c rd : rds) {
+            rd.npe03__Next_Payment_Date__c = getNextDonationDate(schedulesByRdId.get(rd.Id));
+        }
+    }
+
+    /***
+    * @description Retrieves recurring donation schedules associated with recurring donation list
+    * @param rds List<npe03__Recurring_Donation__c>
+    */
+    public Map<Id, List<RecurringDonationSchedule__c>> getExistingSchedules(List<npe03__Recurring_Donation__c> rds) {
+
+        Map<Id, List<RecurringDonationSchedule__c>> oldSchedules = new Map<Id, List<RecurringDonationSchedule__c>>();
+
+        for (RecurringDonationSchedule__c oldSchedule : [
+                                SELECT RecurringDonation__c,
+                                    StartDate__c,
+                                    EndDate__c,
+                                    InstallmentFrequency__c,
+                                    InstallmentPeriod__c,
+                                    DayOfMonth__c
+                                FROM RecurringDonationSchedule__c
+                                WHERE RecurringDonation__c IN :rds
+                                AND Active__c = true
+        ]) {
+            if (!oldSchedules.containsKey(oldSchedule.RecurringDonation__c)) {
+                oldSchedules.put(oldSchedule.RecurringDonation__c, new List<RecurringDonationSchedule__c>());
+            }
+            oldSchedules.get(oldSchedule.RecurringDonation__c).add(oldSchedule);
+        }
+
+        return oldSchedules;
+    }
+
+    /***
+    * @description Adds recurring donation schedules to map by recurring donation Id
+    * @param rd SObject
+    * @param newSchedulesByRdId Map<Id, List<RecurringDonationSchedule__c>>
+    */
+    public void addUpdatedSchedules(RecurringDonationSchedule__c rd, Map<Id, List<RecurringDonationSchedule__c>> newSchedulesByRdId) {
+
+        if (rd.EndDate__c == null || (rd.EndDate__c >= Date.today() && rd.StartDate__c < rd.EndDate__c)) {
+
+            if (!newSchedulesByRdId.containsKey(rd.RecurringDonation__c)) {
+                newSchedulesByRdId.put(rd.RecurringDonation__c, new List<RecurringDonationSchedule__c>());
+            }
+            newSchedulesByRdId.get(rd.RecurringDonation__c).add(rd);
+        }
+    }
+
+    /***
+    * @description Filters recurring donations to determine which need schedule updates
+    * @param rds Trigger.new records
+    * @param oldRds Trigger.old records
+    */
+    public Map<Id, npe03__Recurring_Donation__c> getRecurringDonationsNeedingScheduleUpdates(
```

Comment:
This method is really just checking for changes to the fields, which is more related to the trigger. Is it feasible to have this method in the TDTM class or is there something unique to what it does that it belongs in the service class?

Reply:
It gets called from the methods we just moved.  Do you really want the service to make a call to the trigger method?

Reply:
Possibilities...  Keep this in service or move setScheduleDependentFields() back to trigger.

Reply:
my vote is to move setScheduleDependentFields(), but leave addUpdatedSchedules() and getExistingSchedules() in the service.

Reply:
I've pushed changes that reflect the way I think it should be.  There is no perfect solution here, but I think this comes the closed to what you are trying to achieve without a bunch of gymnastics.

Reply:
I see. How about this?

In the TDTM class:
```
private void calculateNextDonationDatesBeforeUpdate(rds, oldRds) {
    Map<Id, npe03__Recurring_Donation__c> filteredRecurringDonations = getRecurringDonationsNeedingScheduleUpdates(rds, oldRds, TDTM_Runnable.Action.BeforeUpdate);

    Map<Id, List<RecurringDonationSchedule__c>> schedulesByRdId = scheduleService.getSchedules(filteredRecurringDonations)

    for (npe03__Recurring_Donation__c rd : rds) {
        if (schedulesById.containsKey(rd.Id) {
            rd.npe03__Next_Payment_Date__c = getNextDonationDate(schedulesByRdId.get(rd.Id));
        }
    }

}
```

In the ScheduleService class:
```
// Mostly the method that was setScheduleDependentFields without the bits that abstracted into the above
public Map<Id, List<RecurringDonationSchedule__c>> buildSchedules(rds) {

}
```